// sysmsg/index.js
'use strict';

const map = new WeakMap;
const sysmsgs = require('tera-data').sysmsg;

function parseSysmsg(str) {
  if (str[0] !== '@') return null;

  const args = str.slice(1).split('\x0B');
  const id = args.shift();
  const params = {};
  while (args.length > 0) {
    params[args.shift()] = args.shift();
  }

  return { id: id, params: params };
}

function sysmsg(dispatch) {
  this.dispatch = dispatch;
  this.sysmsgs = sysmsgs;

  if (!map.has(dispatch)) {
    map.set(dispatch, {});

    dispatch.hook('sSystemMessage', function sSystemMessage(event) {
      const data = parseSysmsg(event.message);
      if (!data) return;

      const hooks = map.get(dispatch)[data.id];
      if (!hooks) return;

      for (let hook of hooks) {
        const result = hook(data.params);

        // TODO handle other cases
        if (result === false) return false;
      }
    });
  }
};

sysmsg.parse = sysmsg.prototype.parse = parseSysmsg;

sysmsg.prototype.on = function on(msg, cb) {
  if (!(msg in this.sysmsgs.map.name)) {
    console.warn('[sysmsg] Unknown system message "' + msg + '"');
    return;
  }

  const id = this.sysmsgs.map.name[msg];
  const hooks = map.get(this.dispatch);
  if (!hooks[id]) hooks[id] = [];
  hooks[id].push(cb);
};

module.exports = sysmsg;
